soda/contracts/design_decisions.md

```diff
@@ -0,0 +1,52 @@
+# Soda data contract YAML format
+
+### No data source & connection details
+
+We leave datasource, connection and schema out of the contract files
+  * Pro
+    * The same contract file can be used to run in CI/CD
+    * Leveraging the existing Soda configuration files
+  * Con: 
+    * more CLI args
+    * less repeatable
+    * less sharing of the sharable connection details & profiles  
+
+### Column names as keys
+
+* Nicer YAML. A list of columns with a `name` key is less readable.
+* Still JSON Schema enforcable YAML 
+
+### Keys without spaces nor variables 
+
+No spaces in keys.  No parsing of keys.  No variable parts in keys except for the column names.
+
+* Pro: 
+  * More JSON compliant
+  * More validation from JSON schema
+  * More expected and in line with people's expectations
+* Con:
+  * Not similar to SodaCL
+
+# Implementation
+
+### YAML string to YAML string conversion 
+
+We translate Soda data contract YAML format to SodaCL YAML string first and then feed the SodaCL YAML 
+string into a Soda scan.
+
+Pros:
+* Composable.  
+* Easier to build.
+* Users can review the intermediate SodaCL and debug that based on the SodaCL docs.
+
+Cons: 
+* No native error messages on the contract YAML lines.
```

Comment:
- is there any way in which we could somehow solve this technically (not that we **have** to but just thinking outloud.
- do we think we'd ever actually want to maintain a separate parser (like a real parser) so that we can go direct from the contract format into the structures that core then uses, giving us back more control and less compilations. 

just thinking outloud as a "might it be a good idea in the future" kind of thing.

Reply:
If I can chime in from a support perspective, I am understanding data contracts as kind of a layer on top of SodaCL so that contracts get "translated" to SodaCL, then SodaCL gets "translated" to SQL queries. If that approaches a reasonable description of things, then errors for contracts ought to be distinguishable from SodaCL errors so as to be able to troubleshoot. I.e. If I make a syntax error or enter an invalid value in data contract form and I get a None Type error or something like that, is it because Soda "mistranslated" the contract into SodaCL, or is it because I made a typo in my contract?